### PR TITLE
test multiple events emitted across contracts

### PR DESCRIPTION
**Pull request by jobez**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build-related changes
- Documentation content changes
- Testing ☑️ 
- Other (please describe):

## What is the current behavior?

Resolves: #638 

## What is the new behavior?

Adding test that reproduces an issue the kakarot deployment process is seeing
It expects all events to be available in the transaction receipt, whereas the actual case is less. Will circle back and update how many events are present compared to the expected

-
-
-

## Does this introduce a breaking change?

No

### CODE REVIEW

File: cairo-contracts/build/emit_multiple_single_event.json
Comment (ClementWalter) [inline cairo-contracts/build/emit_multiple_single_event.json:1]: the source file seems missing

File: cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo
Comment (ClementWalter) [inline cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo:1]: naming of the events and functions are confusing. It seems like we test internal VS external (or inner VS outer) events, can we try to standardize this?

File: crates/node/src/constants.rs [hunk 1]
```diff
@@ -14,4 +14,10 @@ pub const ERC721_CLASS_HASH: &str = "0x80000";
 pub const UDC_CONTRACT_ADDRESS: &str = "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf";
 pub const UDC_CLASS_HASH: &str = "0x90000";
 
+pub const SINGLE_EVENT_EMITTING_CONTRACT_ADDRESS: &str = "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf";
+pub const SINGLE_EVENT_EMITTING_CLASS_HASH: &str = "0x91000";
+
+pub const MULTIPLE_EVENT_EMITTING_CONTRACT_ADDRESS: &str = "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf";
```
Comment (ClementWalter) [inline crates/node/src/constants.rs:20]: the two addresses are the same, does it really work?

File: crates/pallets/starknet/src/lib.rs [hunk 2]
```diff
@@ -465,6 +465,7 @@ pub mod pallet {
                     }
                 }
                 Err(e) => {
+                    // dbg!(&e);
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/lib.rs:468]: remove or uncomment

File: crates/pallets/starknet/src/tests/call_contract.rs [hunk 3]
```diff
@@ -90,3 +91,27 @@ fn given_call_contract_call_works() {
         pretty_assertions::assert_eq!(res.unwrap(), vec![Felt252Wrapper::from_hex_be("0x02").unwrap()]);
     });
 }
+
+#[test]
+fn given_call_contract_call_works_jhnn() {
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/tests/call_contract.rs:96]: why adding this test in `call_contract` ? I don't think that it emits events

File: crates/pallets/starknet/src/tests/call_contract.rs [hunk 4]
```diff
@@ -1,10 +1,11 @@
+use blockifier::abi::abi_utils::selector_from_name;
 use frame_support::{assert_ok, bounded_vec};
 use mp_starknet::execution::types::Felt252Wrapper;
 use mp_starknet::transaction::types::InvokeTransaction;
 use sp_core::ConstU32;
 use sp_runtime::BoundedVec;
 
-use super::constants::TOKEN_CONTRACT_CLASS_HASH;
+use super::constants::{MULTIPLE_EVENT_EMITTING_CONTRACT_ADDRESS, TOKEN_CONTRACT_CLASS_HASH};
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/tests/call_contract.rs:8]: you don't use the other contract?

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 5]
```diff
@@ -203,6 +203,76 @@ fn given_hardcoded_contract_run_invoke_tx_then_event_is_emitted() {
...
 }
 
+#[test]
+fn given_hardcoded_contract_run_invoke_tx_then_multiple_events_is_emitted() {
+    new_test_ext().execute_with(|| {
+        System::set_block_number(0);
+        run_to_block(2);
+
+        let emit_contract_address =
+            Felt252Wrapper::from_hex_be("0x051a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf").unwrap();
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:213]: why don't you import it from constants?

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 6]
```diff
@@ -203,6 +203,76 @@ fn given_hardcoded_contract_run_invoke_tx_then_event_is_emitted() {
...
+        let sender_account = get_account_address(AccountType::NoValidate);
+
+        let events = System::events();
+
+        let do_one_event_transaction = InvokeTransaction {
+            version: 1,
+            sender_address: sender_account,
+            calldata: bounded_vec![
+                emit_contract_address, // Token address
+                Felt252Wrapper::from_hex_be("0x1f9630afc69718ef80a0353d14ae7e8e4003b0f4def439cfdde1843fdff1833")
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:224]: why don't you use `get_selector_from_name` as in the `call_contract` example?

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 7]
```diff
@@ -203,6 +203,76 @@ fn given_hardcoded_contract_run_invoke_tx_then_event_is_emitted() {
...
+            ],
+            nonce: Felt252Wrapper::ONE,
+            max_fee: Felt252Wrapper::from(u128::MAX),
+            signature: bounded_vec!(),
+        };
+
+        let none_origin = RuntimeOrigin::none();
+
+        let chain_id = Starknet::chain_id().0.to_bytes_be();
+        let chain_id = std::str::from_utf8(&chain_id[..]).unwrap();
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:262]: you already defined `chain_id` previously

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 8]
```diff
@@ -203,6 +203,76 @@ fn given_hardcoded_contract_run_invoke_tx_then_event_is_emitted() {
...
+        let transaction_hash = calculate_invoke_tx_hash(do_one_event_transaction.clone(), chain_id);
+
+        // dbg!(&do_one_event_transaction);
+        Starknet::invoke(none_origin, do_one_event_transaction).expect("work");
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:240]: is there any expect of events content?

File: crates/pallets/starknet/src/tests/invoke_tx.rs [hunk 9]
```diff
@@ -203,6 +203,76 @@ fn given_hardcoded_contract_run_invoke_tx_then_event_is_emitted() {
...
+        let transaction_hash = calculate_invoke_tx_hash(do_two_event_transaction.clone(), chain_id);
+
+        // dbg!(&do_two_event_transaction);
+        Starknet::invoke(none_origin, do_two_event_transaction).expect("work");
+
+        let pending = Starknet::pending();
+        let two_receipt = &pending.get(1).unwrap().1;
+        // dbg!(&pending);
+        // dbg!(&two_receipt);
+        assert!(one_receipt.events.len() < two_receipt.events.len(), "Receipt1 has more or equal events than Receipt2");
```
Comment (ClementWalter) [inline crates/pallets/starknet/src/tests/invoke_tx.rs:272]: can't we just assert the expected content? why this relative assert?

File: crates/primitives/starknet/src/execution/call_entrypoint_wrapper.rs [hunk 10]
```diff
@@ -92,7 +92,11 @@ impl CallEntryPointWrapper {
 
         call_entry_point
             .execute(state, execution_resources, execution_context, &block_context, &account_context)
-            .map_err(EntryPointExecutionErrorWrapper::EntryPointExecution)
+            .map_err(|err| {
```
Comment (ClementWalter) [inline crates/primitives/starknet/src/execution/call_entrypoint_wrapper.rs:95]: undo plz

**Review by ClementWalter — CHANGES_REQUESTED** [review]

File: crates/node/src/constants.rs (hunk 1 above)
Comment (jobez) [reply crates/node/src/constants.rs:20]: this is the node constants, not the mock. i erroneously added. will remove

File: crates/pallets/starknet/src/tests/call_contract.rs (hunk 3 above)
Comment (jobez) [reply crates/pallets/starknet/src/tests/call_contract.rs:96]: this was me building understanding that i accurately seeded the mock with a contract

File: crates/pallets/starknet/src/tests/invoke_tx.rs (hunk 6 above)
Comment (jobez) [reply crates/pallets/starknet/src/tests/invoke_tx.rs:224]: i believe they return different types

File: crates/pallets/starknet/src/tests/call_contract.rs (hunk 4 above)
Comment (jobez) [reply crates/pallets/starknet/src/tests/call_contract.rs]: the other contract is initialized in a storage var of this contract as a way to test external event emission

**Review by ClementWalter — APPROVED** [review]
thx! But the bug is not visible then right?

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

**Review by LucasLvy — APPROVED** [review]

**Pull request merged**